npm/src/language-service/language-service.ts

```diff
@@ -130,28 +161,29 @@ export class QSharpLanguageService implements ILanguageService {
     for (const cell of cells) {
       this.code[cell.uri] = cell.code;
     }
-    this.languageService.update_notebook_document(notebookUri, cells);
+    await this.languageService.update_notebook_document(notebookUri, cells);
   }
 
   async closeDocument(documentUri: string): Promise<void> {
     delete this.code[documentUri];
-    this.languageService.close_document(documentUri);
+    await this.languageService.close_document(documentUri);
   }
 
   async closeNotebookDocument(
     documentUri: string,
     cellUris: string[],
   ): Promise<void> {
     cellUris.forEach((uri) => delete this.code[uri]);
-    this.languageService.close_notebook_document(documentUri, cellUris);
+    await this.languageService.close_notebook_document(documentUri, cellUris);
   }
 
   async getCompletions(
     documentUri: string,
     offset: number,
   ): Promise<ICompletionList> {
-    const code = this.code[documentUri];
-    if (code === undefined) {
+    const code = await this.loadFile(documentUri);
```

Comment:
Was there a specific reason to change these to file reads? I don't see why it should be necessary - the source code contents should have already been populated in a prior `updateDocument` call. I'd prefer to go back to the existing map lookup. 


You can, however, add these `loadFile` calls to https://github.com/microsoft/qsharp/blob/17cc2d253522f0d3d57f5b51f1a21ac89c414823/npm/src/language-service/language-service.ts#L260-L264 and similar locations in `getReferences` and `getRename`, since those are the points where we're about to navigate to an *unopened* file and we need the file contents.

Reply:
Oh, I just noticed `loadFile` does the map lookup first! I think that's alright, but then you'll want to update the remaining spots in this file that do the `this.code[uri]` lookups. You missed a few (probably added after you made the changes)...